pkg/smb/nodeserver_test.go

```diff
@@ -198,18 +210,20 @@ func TestNodeExpandVolume(t *testing.T) {
 }
 
 func TestNodePublishVolume(t *testing.T) {
-	skipIfTestingOnWindows(t)
 	volumeCap := csi.VolumeCapability_AccessMode{Mode: csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER}
-	errorMountSource := "./error_mount_source"
-	alreadyMountedTarget := "./false_is_likely_exist_target"
-	smbFile := "./smb.go"
-	sourceTest := "./source_test"
-	targetTest := "./target_test"
+
+	errorMountSource := getTestPath("error_mount_source")
+	alreadyMountedTarget := getTestPath("false_is_likely_exist_target")
+	smbFile := getTestPath("smb.go")
+	sourceTest := getTestPath("source_test")
+	targetTest := getTestPath("target_test")
 
 	tests := []struct {
-		desc        string
-		req         csi.NodePublishVolumeRequest
-		expectedErr error
+		desc               string
+		req                csi.NodePublishVolumeRequest
+		expectedErr        error
+		expectedErrUnix    error
+		expectedErrWindows error
```

Comment:
We currently have code to differentiate Windows and Linux errors. Try taking a look at - https://github.com/kubernetes-csi/csi-driver-smb/blob/master/test/utils/testutil/testutil.go#L28

This object holds the error messages for both Windows and Linux as a single entity. Also try looking at the following helper methods for assertion. 
Why don't we try reusing that instead of re-introducing the same funcitonility?

Here's an example of this - https://github.com/kubernetes-csi/csi-driver-smb/blob/master/pkg/smb/nodeserver_test.go#L60

Reply:
Thanks will update the PR accordingly.

Reply:
I think using we might have to make some modification to the existing struct: 
```
type TestError struct {
	WindowsError error
	DefaultError error
}
```
In this struct, we are taking the DefaultError when the WindowsError is `nil`. But there are cases where we are explicitly asking to check with nil. In such cases its bit difficult to check for errors. So I propose a small modification to the struct to have a fields like:
```
type TestError struct {
        expectedError error
	WindowsError error
	DefaultError error
}
So that explictly checking for windows error even if its nil is easier.

what are your thoughts on this?

Reply:
@boddumanohar We're actually following the current approach in the azurefile and azuredisk repos for unit testing - https://github.com/kubernetes-sigs/azurefile-csi-driver , https://github.com/kubernetes-sigs/azuredisk-csi-driver

We didn't really have to make changes to the assertion logic there. It seems like you're suggesting this change because of the unexpected error given by `TestNodeUnpublishVolume` (`[Error] Unmount error mocked by IsLikelyNotMountPoint`) and `TestNodeUnstageVolume` (`[Error] CleanupMountPoint error mocked by IsLikelyNotMountPoint`). IMO, because these are unexpected errors from the CSI proxy, its best if we temporarily skip only these particular test cases on Windows until we figure out why CSI proxy is giving that issue, rather that changing the assertion logic to incorporate this unexpected failures.

I've opened a PR here with the above thoughts, and keeping the same approach as azuredisk and azurefile repos - https://github.com/kubernetes-csi/csi-driver-smb/pull/123

@boddumanohar @andyzhangx Please let me know your thoughts, and we can make more changes accordingly. Thanks! :)

Reply:
I think that's case by case
 - `error mocked by IsLikelyNotMountPoint` error is likely caused by no fake of `IsLikelyNotMountPoint` on Linux, so need to fix it on Linux, that may fix the issue?
- `[Error] Mount error mocked by Mount`
  - it's caused by different mount behavior between linux and windows, for this case, let's only run on Linux

In general, I think we should use https://github.com/kubernetes-csi/csi-driver-smb/blob/master/test/utils/testutil/testutil.go#L28, it's more clear.

And if there is one case that only Linux would return error(Windows return nil), we could make it separate, e.g. add `OnlyRunOnPlatform` in test case struct, that would make logic more clear: positive and negative cases should be separated

Reply:
point is `TestError` should really contains error, if it's nil, then it's not error. if there is one case that only Linux would return error(Windows return nil), we could make it separate, e.g. add `OnlyRunOnPlatform` in test case struct, that would make logic more clear: positive and negative cases should be separated

Reply:
Cool. Let me try this out.